src/main/java/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataEx3Dao.java

```diff
@@ -0,0 +1,133 @@
+package org.sagebionetworks.bridge.dynamodb;
+
+import java.util.List;
+import java.util.Optional;
+
+import javax.annotation.Resource;
+
+import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapper;
+import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBQueryExpression;
+import com.amazonaws.services.dynamodbv2.model.AttributeValue;
+import com.amazonaws.services.dynamodbv2.model.ComparisonOperator;
+import com.amazonaws.services.dynamodbv2.model.Condition;
+import com.google.common.collect.ImmutableList;
+import org.springframework.stereotype.Component;
+
+import org.sagebionetworks.bridge.BridgeUtils;
+import org.sagebionetworks.bridge.dao.HealthDataEx3Dao;
+import org.sagebionetworks.bridge.models.healthdata.HealthDataRecordEx3;
+
+@Component
+public class DynamoHealthDataEx3Dao implements HealthDataEx3Dao {
+    private DynamoDBMapper mapper;
+
+    @Resource(name = "healthDataEx3DdbMapper")
+    public void setMapper(DynamoDBMapper mapper) {
+        this.mapper = mapper;
+    }
+
+    @Override
+    public HealthDataRecordEx3 createOrUpdateRecord(HealthDataRecordEx3 record) {
+        DynamoHealthDataRecordEx3 dynamoRecord = (DynamoHealthDataRecordEx3) record;
+
+        if (dynamoRecord.getId() == null) {
+            // This record doesn't have its ID assigned yet (new record). Create an ID and assign it.
+            dynamoRecord.setId(BridgeUtils.generateGuid());
+        }
+
+        // Save to DDB.
+        mapper.save(dynamoRecord);
+        return dynamoRecord;
+    }
+
+    @Override
+    public void deleteRecordsForHealthCode(String healthCode) {
+        // First, query the records we need to delete.
+        DynamoHealthDataRecordEx3 key = new DynamoHealthDataRecordEx3();
+        key.setHealthCode(healthCode);
+
+        DynamoDBQueryExpression<DynamoHealthDataRecordEx3> query = new DynamoDBQueryExpression<DynamoHealthDataRecordEx3>()
+                .withConsistentRead(false).withIndexName(DynamoHealthDataRecordEx3.HEALTHCODE_CREATEDON_INDEX)
+                .withHashKeyValues(key);
+        List<DynamoHealthDataRecordEx3> recordsToDelete = queryHelper(query);
```

Comment:
Doesn't the query "time out" at 1MB of data returned...what happens if the user has more data than this?

Reply:
The 1MB limitation is on queryPage(), which gets one page of the data. The max page size is 1MB. query() automatically handles pagination for us.